pilot/pkg/security/authz/policy/helper.go

```diff
@@ -229,3 +237,32 @@ func Verify(got *envoy_rbac.RBAC, want map[string][]string) error {
 
 	return err
 }
+
+// checkSourcePrincipalCounts checks if |actual| and |want| have the same number of source.principal.
+// For example, both |actual| and |want| can only have one key "source.principal",
+// but |actual| can have "foo" and "bar", and |want| can only has "foo". This is possible since
+// in v1beta1, principals can be a list.
+func checkPrincipalCounts(ruleKey string, actual string, want []string) error {
+	// Only work with principals, not other identities like namespaces.
+	keyRegEx := regexp.MustCompile("source.principal")
```

Comment:
This is hard to maintain and could easily cause false positive/negative, can we actually check the appearance of the principal in `want`?

Reply:
Currently, `want` outputs something like this `[MethodFromRole[role] -td/ns/foo/sa/bar td1/ns/foo/sa/bar td2/ns/foo/sa/bar]`. Would it be better if we add some prefix tag before the principal, like `Principal[-td/ns/foo/sa/bar]`? So that we can easily extract the principals out?

Reply:
Why is it hard to maintain? Previously we were checking the principal from `want`, but something `actualStr` may have more principals than `want`, which is a false positive. For example. want = [foo], actualStr = "....source.principal: foo....source.principal: bar".